File: leetcode/cpp/myHeapq.cpp
// 个人实现的小顶堆

#include <iostream>
#include <vector>
#include <assert.h>

template <class T>
void swap(std::vector<T>& vec, int a, int b) {
    T tmp = vec[a];
    vec[a] = vec[b];
    vec[b] = tmp;
}

template <class T>
void shift_up(std::vector<T>& vec, int i) {
    while (1) {
        int p = (i - 1) / 2;
        if (p >= 0 && vec[p] > vec[i]) {
            swap(vec, i, p);
            i = p;
        } else {
            break;
        }
    }
}

template <class T>
void shift_down(std::vector<T>& vec, int n, int i) {
    while (1) {
        int smallest = i;
        int l = 2 * i + 1;
        int r = 2 * i + 2;
        if (l < n && vec[l] < vec[smallest]) {
            smallest = l;
        }
        if (r < n && vec[r] < vec[smallest]) {
            smallest = r;
        }
        if (smallest != i) {
            swap(vec, i, smallest);
            i = smallest;
        } else {
            break;
        }
    }
}

template <class T>
void heapify(std::vector<T>& vec) {
    assert(!vec.empty());
    for (int i = vec.size()/2-1; i >= 0; i--) {
        shift_down(vec, vec.size(), i);
    }
}

template <class T>
T heappop(std::vector<T>& vec) {
    assert(!vec.empty());
    swap(vec, 0, vec.size() - 1);
    shift_down(vec, vec.size(), 0);
    int ret = vec.back();
    vec.pop_back();
    return ret;
}

template <class T>
void heappush(std::vector<T>& vec, T num) {
    vec.push_back(num);
    shift_up(vec, vec.size() - 1);
}

template <class T>
void printf_1d_vec(const std::vector<T>& vec) {
    for (auto & num : vec) {
        std::cout << num << ", ";
    }
    std::cout << std::endl;
}

int main(int argc, char* argv[])
{
    std::vector<int> vec = {5, 2, 6, 2, 0, 5, 9, 12, 3};
    printf_1d_vec(vec);
    heapify(vec);
    printf_1d_vec(vec);
    heappop(vec);
    printf_1d_vec(vec);
    heappush(vec, -1);
    printf_1d_vec(vec);
    return 0;
}

File: leetcode/python/myHeapq.py
#!/bin/bash python
"""
个人实现的小顶堆
"""

def shift_up(vec, i):
    """
    索引为i的元素上浮到合适位置
    """
    while True:
        p = (i - 1) // 2
        if p >= 0 and vec[p] > vec[i]:
            vec[i], vec[p] = vec[p], vec[i]
            i = p
        else:
            break


def shift_down(vec, n, i):
    """
    索引为i的元素下沉到合适位置
    """
    while True:
        smallest = i
        l = 2 * i + 1
        r = 2 * i + 2
        if l < n and vec[l] < vec[smallest]:
            smallest = l
        if r < n and vec[r] < vec[smallest]:
            smallest = r
        if smallest != i:
            vec[i], vec[smallest] = vec[smallest], vec[i]
            i = smallest
        else:
            break


def heapify(vec):
    """
    原地建堆
    """
    assert vec != []
    n = len(vec)
    for i in range(n//2-1, -1, -1):
        shift_down(vec, n, i)


def heappop(vec):
    """
    弹出最小值
    """
    assert vec != []
    n = len(vec)
    vec[0], vec[n - 1] = vec[n - 1], vec[0] # 交换堆首堆尾
    shift_down(vec, n - 1, 0) # 堆首下沉到合适位置
    return vec.pop()


def heappush(vec, num):
    """
    压入新元素
    """
    vec.append(num) # 堆尾添加新元素
    shift_up(vec, len(vec) - 1) # 堆尾上浮


if __name__ == "__main__":
    vec = [5, 2, 6, 2, 0, 5, 9, 12, 3]
    print(vec)
    heapify(vec)
    print(vec)
    heappop(vec)
    print(vec)
    heappush(vec, -1)
    print(vec)